### PR TITLE
Build clrcompression.dll with optimized Zlib-Intel

### PR DESCRIPTION
**Pull request by bjjones**
#3986 Zlib-Intel is an improved version of Zlib that contains optimizations for processors with the SSE4.2 instruction set. In the most common case of CompressionLevel Optimal, we see a +25-30% performance improvement. 

Currently, the version of clrcompression.dll built in the repo is not copied into any of the tests, so to test I manually copied the binary into System.IO.Compression.Tests and ran the test manually resulting in no errors.

Please note that due to a gzip incompatibility I have modified the library to no longer use the agressive deflate_quick strategy and instead use the traditional deflate_fast strategy for CompressionLevel Fastest. This causes performance for CompressionLevel Fastest to be similar to Zlib-Adler.

For ARM platforms, clrcompression.dll is still built using Zlib-Adler.

Performance numbers gathered on an Intel Core i5-4670 CPU for the Canterbury Corpus Benchmark follow: 

| innerIterations  | filename | compressLevel | Adler | Intel | Intel / Adler |
| :------------ |:---------------|:-----|:-----|:-----|:-----|
| 25 | alice29.txt |Optimal|11.12274|7.505396|67.48%|
| 25 | asyoulik.txt |  Optimal |10.22775|6.969144|68.14%|
| 25 | cp.html |   Optimal |1.28132|1.4261|110.30%|
| 25 | fields.c  |   Optimal |0.81646|0.8408881|102.99%|
| 25 | grammar.lsp       |   Optimal |0.7443761|0.670172|90.03%|
| 25 | kennedy.xls   |  Optimal |48.62418|18.24115|37.51%|
| 25 | lcet10.txt      |    Optimal|26.82975|17.67865|65.89%|
| 25 | plrabn12.txt       |    Optimal |39.65364|25.09914|63.3%|
| 25 | ptt5   |    Optimal |14.31153|9.271148|64.78%|
| 25 | sum        |    Optimal |10.1662|3.163316|31.12%|
| 25 | xargs.1      |    Optimal |0.831952|0.7740601|93.04%|

| innerIterations  | CompressLevel | Average Intel / Adler |
| :------------ |:---------------|:-----|
| 1 | Optimal |75.69%|
| 10 | Optimal |70.62%|
| 25 | Optimal |70.51%|
| 50 | Optimal |71.75%|
| 1 | Fastest |101.53%|
| 10 | Fastest |100.38%|
| 25 | Fastest |95.73%|
| 50 | Fastest |99.54%|
| 1 | NoCompression |100.07%|
| 10 | NoCompression |100.78%|
| 25 | NoCompression |97.40%|
| 50 | NoCompression |99.26%|

@ianhays @stephentoub @joshfree

### CODE REVIEW

**Comment by stephentoub** [comment]
Thanks, @bjjones.  Does this / how does this affect compression levels / output file sizes?

**Comment by bjjones** [comment]
For the Canterbury Corpus at CompressionLevel Optimal using deflateStream, I saw:

| Library  | Initial Size |Compressed Size|Ratio (Compressed/Initial) | 
| :------------ |:---------------|:-----|:-----|
|Zlib-Adler|2670KB| 713KB | .26704|
|Zlib-Intel|2670KB| 733KB | .27453 |

Also noted in the Zlib-Intel Whitepaper, there are very small increases in compression ratio: http://www.intel.com/content/dam/www/public/us/en/documents/white-papers/zlib-compression-whitepaper-copy.pdf

**Comment by stephentoub** [comment]
> For the Canterbury Corpus at CompressionLevel Optimal using deflateStream

Thanks for the data.  What about at Fastest?

**Comment by bjjones** [comment]
In the same scenario, I saw a slightly improved compression ratio using CompressionLevel Fastest.

| Library  | Initial Size  | Compressed Size | Ratio (Compressed/Initial) |
| :------------ |:---------------| :-----| :-----|
| Zlib-Adler | 2670KB  | 842KB | .31535|
| Zlib-Intel   | 2670KB       |   812KB | .30412 |

**Comment by ianhays** [comment]
How many of the zlib-intel files are different than their zlib-adler equivalents? For those that are the same, could we just modify the CMakeLists for zlib-intel to reference the ones in zlib-adler? So we don't have the duplication of identical files and it's also easier to see what zlib-intel changes.

It would mess up the MakeFile build, but we don't use that anyways.

File: src/Native/Windows/build-native.cmd [hunk 1]
```diff
@@ -9,7 +9,7 @@ set __CMakeBinDir=""
 set __IntermediatesDir=""
 set __BuildArch=AnyCPU
 set __VCBuildArch=x86_amd64
-set CMAKE_BUILD_TYPE=Debug
+set CMAKE_BUILD_TYPE=Release
```
Comment (ianhays) [inline src/Native/Windows/build-native.cmd]: good call

File: src/Native/Windows/clrcompression/CMakeLists.txt [hunk 2]
```diff
@@ -7,6 +7,33 @@ endif()
 set(CMAKE_INCLUDE_CURRENT_DIR ON)
 include (GenerateExportHeader)
 
+if($ENV{__BuildArch} STREQUAL x86 OR $ENV{__BuildArch} STREQUAL x64)
+set(NATIVECOMPRESSION_SOURCES
+    clrcompression.def
+    zlib-intel/adler32.c
```
Comment (ianhays) [inline src/Native/Windows/clrcompression/CMakeLists.txt]: e.g. if zlib-intel/adler32.c == zlib/adler32.c then modify these sources to use zlib/adler32.c

File: src/Native/Windows/clrcompression/zlib-intel/README [hunk 3]
```diff
@@ -0,0 +1,115 @@
+ZLIB DATA COMPRESSION LIBRARY
```
Comment (ianhays) [inline src/Native/Windows/clrcompression/zlib-intel/README]: Should we add something in the README, FAQ, ChangeLog, Index, etc. about zlib-intel and the changes that it brings?

File: src/Native/Windows/clrcompression/CMakeLists.txt (hunk 2 above)
Comment (ianhays) [inline src/Native/Windows/clrcompression/CMakeLists.txt]: Also, are all of these files used? I know for zlib-adler there were a bunch of files that weren't getting used in the build so we commented out their inclusion. Can we do that here?

**Comment by ianhays** [comment]
I ran the perf&unit tests again on the zlibs produced by the cmake build and replicated your results.

Looks good other than a few minor comments about duplication :)

**Comment by ianhays** [comment]
@mellinoe OSX_Release failure https://github.com/dotnet/corefx/issues/4833 presumably
13:30:18    System.Numerics.Tests.Matrix4x4Tests.Matrix4x4CreateFromYawPitchRollTest2 [FAIL]
13:30:18       Yaw:155 Pitch:-20 Roll:50
13:30:18       Expected: True
13:30:18       Actual:   False
13:30:18       Stack Trace:
13:30:18             at System.Numerics.Tests.Matrix4x4Tests.Matrix4x4CreateFromYawPitchRollTest2()

**Comment by bjjones** [comment]
Addressed feedback and I think we should be good to go.

(I messed up when trying to squash the commits. Sorry for the thrash)

**Comment by bjjones** [comment]
Unnecessary/duplicate files have been removed and I added a README.txt to the Zlib-Intel folder.

**Comment by ianhays** [comment]
LGTM. @stephentoub ?

File: src/Native/Windows/build-native.cmd (hunk 1 above)
Comment (stephentoub) [inline src/Native/Windows/build-native.cmd]: Shouldn't this be configured based on the type of build being done?

File: src/Native/Windows/clrcompression/CMakeLists.txt [hunk 4]
```diff
@@ -7,6 +7,34 @@ endif()
...
+    #zlib/gzlib.c
+    #zlib/gzread.c
+    #zlib/gzwrite.c
+    #zlib/uncompr.c
+    #zlib/infback.c
+    #zlib/uncompr.c
+)
+endif()
+
+if($ENV{__BuildArch} STREQUAL arm)
```
Comment (stephentoub) [inline src/Native/Windows/clrcompression/CMakeLists.txt]: I don't know what other potential future values __BuildArch could/should be, but should this just be an ```else```?

File: src/Native/Windows/build-native.cmd (hunk 1 above)
Comment (ianhays) [inline src/Native/Windows/build-native.cmd]: It is. This is the default setting. It is set on lines 19/20

File: src/Native/Windows/clrcompression/zlib-intel/README.txt [hunk 5]
```diff
@@ -0,0 +1,11 @@
+Zlib-Intel is an optimized version of Zlib developed by Intel to take advantage of modern CPU features.
```
Comment (stephentoub) [inline src/Native/Windows/clrcompression/zlib-intel/README.txt]: Do we need to make any changes to the 3rd party notices file?
https://github.com/dotnet/corefx/blob/master/THIRD-PARTY-NOTICES#L65

cc: @richlander, @martinwoodward

File: src/Native/Windows/build-native.cmd (hunk 1 above)
Comment (stephentoub) [inline src/Native/Windows/build-native.cmd]: I see.  Why change the default?  Don't we normally default to debug unless release is explicitly chosen?

File: src/Native/Windows/clrcompression/zlib-intel/adler32.c [hunk 6]
```diff
@@ -0,0 +1,195 @@
+/* adler32.c -- compute the Adler-32 checksum of a data stream
+ * Copyright (C) 1995-2011 Mark Adler
+ * For conditions of distribution and use, see copyright notice in zlib.h
+ */
+
+/* @(#) $Id$ */
```
Comment (stephentoub) [inline src/Native/Windows/clrcompression/zlib-intel/adler32.c]: What is this ```@(#) $Id$``` comment showing up in the files?

**Comment by stephentoub** [comment]
A few nits, but generally looks good to me.

File: src/Native/Windows/clrcompression/zlib-intel/adler32.c (hunk 6 above)
Comment (ellismg) [inline src/Native/Windows/clrcompression/zlib-intel/adler32.c]: [Other source control systems](http://svnbook.red-bean.com/en/1.4/svn.advanced.props.special.keywords.html) expand that on checkout to something meaningful (e.g. the filename, last modified date, etc).

**Comment by stephentoub** [comment]
Test Innerloop OSX Debug Build and Test please
Test Innerloop OSX Release Build and Test please

File: src/Native/Windows/clrcompression/zlib-intel/README.txt (hunk 5 above)
Comment (martinwoodward) [inline src/Native/Windows/clrcompression/zlib-intel/README.txt]: Just to confirm - the notices file already includes the appropriate ZLib reference due to the code on which this PR is based so we're good as is. Thanks for checking.

**Comment by stephentoub** [comment]
Test Innerloop OSX Release Build and Test please

**Comment by stephentoub** [comment]
Thanks, @bjjones.

**Pull request merged**